### ISSUE TITLE
Pluggable search

### ISSUE THREAD

**Issue by danielballan** [NSLS-II/caproto#452]
Clients should be able to consult a mapping before initiating UDP broadcast. This mapping could be backed by:

* aplain text file of PV names and addresses
* etcd
* channel finder?

Any of these is likely to be significantly faster than UDP broadcast. If that mapping fails to resolve successfully, UDP will still be available as a fallback. On the server side, servers need to be able to update this mapping at startup, shutdown, and at arbitrary points during their runtime (see new dynamic IOC exmaples).

**Comment by klauer** [comment]
Related: we should determine how we're going to keep these "`CA++`"-style additions cleanly separate - but yet still accessible - from the base CA-only implementation.

**Comment by danielballan** [comment]
Yes. I think that question is important for #453. For this, I think clients are always allowed to initiate the TCP phase with a server at some address they happen to know, so this it just surfacing API to make this more obvious. I wouldn't say that this even rises to the level of an extension to the protocol.

**Comment by tacaswell** [comment]
I would weakly argue that this is not an extension of the protocol because clients are not _obligated_ to search before connecting.  We already have a notion of caching the search results with in the client, this is just pushing that cache out / making it fancier.

I would also add using a redis server and consulting it first before doing the broadcast should also be added to the list (expiration via https://redis.io/commands/expire like we do now in-process).

**Comment by klauer** [comment]
Regardless of how you classify it - these fun enhancements/additions should be very optional and out-of-the-way.
